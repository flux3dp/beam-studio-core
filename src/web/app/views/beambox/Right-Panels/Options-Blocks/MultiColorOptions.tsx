import React, { useEffect, useState } from 'react';

import ColorPicker from 'app/widgets/ColorPicker';
import colloectColors from 'helpers/color/collectColors';
import HistoryCommandFactory from 'app/svgedit/HistoryCommandFactory';
import ISVGCanvas from 'interfaces/ISVGCanvas';
import symbolMaker from 'helpers/symbol-maker';
import { getSVGAsync } from 'helpers/svg-editor-helper';

import styles from './MultiColorOptions.module.scss';

let svgCanvas: ISVGCanvas;
getSVGAsync((globalSVG) => {
  svgCanvas = globalSVG.Canvas;
});

interface Props {
  elem: Element;
}

// TODO: add tests
const MultiColorOptions = ({ elem }: Props): JSX.Element => {
  const [colors, setColors] = useState(colloectColors(elem));
  useEffect(() => {
    setColors(colloectColors(elem));
  }, [elem]);

  const handleColorChange = (origColor: string, newColor: string) => {
    const fillElements = [];
    const strokeElements = [];
    colors[origColor].forEach(({ element, attribute }) => {
      if (attribute === 'fill') fillElements.push(element);
      else if (attribute === 'stroke') strokeElements.push(element);
    });
    const batchCmd = HistoryCommandFactory.createBatchCommand('Update Color');
    if (fillElements.length > 0) {
      svgCanvas.undoMgr.beginUndoableChange('fill', fillElements);
      svgCanvas.changeSelectedAttributeNoUndo('fill', newColor, fillElements);
      const cmd = svgCanvas.undoMgr.finishUndoableChange();
      if (!cmd.isEmpty()) batchCmd.addSubCommand(cmd);
    }
    if (strokeElements.length > 0) {
      svgCanvas.undoMgr.beginUndoableChange('stroke', strokeElements);
      svgCanvas.changeSelectedAttributeNoUndo('stroke', newColor, strokeElements);
      const cmd = svgCanvas.undoMgr.finishUndoableChange();
      if (!cmd.isEmpty()) batchCmd.addSubCommand(cmd);
    }
    if (!batchCmd.isEmpty()) svgCanvas.undoMgr.addCommandToHistory(batchCmd);
    setColors(colloectColors(elem));
    if (elem.tagName.toLowerCase() === 'use') {
      symbolMaker.reRenderImageSymbol(elem as SVGUseElement, { force: true });
    }
  };
  // TODO: controls horizontal scroll handler for windows
  return (
    <div className={styles.block}>
      <div className={styles.label}>Colors</div>
      <div className={styles.controls}>
        {Object.keys(colors).map((color) => (
          <ColorPicker
            key={color}
            initColor={color}
            onChange={(newColor) => handleColorChange(color, newColor)}
          />
        ))}
      </div>
    </div>
  );
};

export default MultiColorOptions;
