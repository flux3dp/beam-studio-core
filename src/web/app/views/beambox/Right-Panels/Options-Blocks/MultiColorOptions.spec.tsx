import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';

import MultiColorOptions from './MultiColorOptions';

jest.mock(
  'app/widgets/ColorPicker',
  () =>
    ({ allowClear, initColor, triggerType, onChange }: any) =>
      (
        <div>
          Mock ColorPicker
          <p>allowClear: {allowClear ? 't' : 'f'}</p>
          <p>initColor: {initColor}</p>
          <p>triggerType: {triggerType}</p>
          <button type="button" onClick={() => onChange('#AAAAFF')}>
            onChange
          </button>
        </div>
      )
);

const mockColloectColors = jest.fn();
jest.mock(
  'helpers/color/collectColors',
  () =>
    (...args) =>
      mockColloectColors(...args)
);

const mockCreateBatchCommand = jest.fn();
jest.mock('app/svgedit/HistoryCommandFactory', () => ({
  createBatchCommand: (...args) => mockCreateBatchCommand(...args),
}));

const mockAddSubCommand = jest.fn();
const mockBatchCommand = {
  addSubCommand: (...args) => mockAddSubCommand(...args),
  isEmpty: () => false,
};

const mockReRenderImageSymbol = jest.fn();
jest.mock('helpers/symbol-maker', () => ({
  reRenderImageSymbol: (...args) => mockReRenderImageSymbol(...args),
}));

const mockChangeSelectedAttributeNoUndo = jest.fn();
const mockBeginUndoableChange = jest.fn();
const mockFinishUndoableChange = jest.fn();
const mockAddCommandToHistory = jest.fn();
jest.mock('helpers/svg-editor-helper', () => ({
  getSVGAsync: (callback) => {
    callback({
      Canvas: {
        changeSelectedAttributeNoUndo: (...args) => mockChangeSelectedAttributeNoUndo(...args),
        undoMgr: {
          beginUndoableChange: (...args) => mockBeginUndoableChange(...args),
          finishUndoableChange: (...args) => mockFinishUndoableChange(...args),
          addCommandToHistory: (...args) => mockAddCommandToHistory(...args),
        },
      },
    });
  },
}));

jest.mock('app/widgets/HorizontalScrollContainer', () => ({ className, children }: any) => (
  <div className={className}>{children}</div>
));

describe('test MultiColorOptions', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    mockColloectColors.mockReturnValue({
      '#AAFFFF': [{ element: '1', attribute: 'fill' }],
      '#FFAAFF': [
        { element: '2', attribute: 'fill' },
        { element: '3', attribute: 'stroke' },
      ],
    });
    mockCreateBatchCommand.mockReturnValue(mockBatchCommand);
  });

  it('should render correctly', () => {
    const { container } = render(<MultiColorOptions elem={document.createElement('rect')} />);
    expect(container).toMatchSnapshot();
  });

  test('editing color in use element should work', async () => {
    const { getByText, getAllByText } = render(
      <MultiColorOptions elem={document.createElement('use')} />
    );
    const mockChangeCmd = { isEmpty: () => false };
    mockColloectColors.mockReturnValue({
      '#AAFFFF': [{ element: '1', attribute: 'fill' }],
      '#AAAAFF': [
        { element: '2', attribute: 'fill' },
        { element: '3', attribute: 'stroke' },
      ],
    });
    mockFinishUndoableChange.mockReturnValue(mockChangeCmd);
    await act(async () => {
      fireEvent.click(getAllByText('onChange')[1]);
    });
    expect(mockCreateBatchCommand).toBeCalledTimes(1);
    expect(mockCreateBatchCommand).toHaveBeenLastCalledWith('Update Color');
    expect(mockBeginUndoableChange).toBeCalledTimes(2);
    expect(mockBeginUndoableChange).toHaveBeenNthCalledWith(1, 'fill', ['2']);
    expect(mockBeginUndoableChange).toHaveBeenNthCalledWith(2, 'stroke', ['3']);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenCalledTimes(2);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenNthCalledWith(1, 'fill', '#AAAAFF', ['2']);
    expect(mockChangeSelectedAttributeNoUndo).toHaveBeenNthCalledWith(2, 'stroke', '#AAAAFF', [
      '3',
    ]);
    expect(mockFinishUndoableChange).toHaveBeenCalledTimes(2);
    expect(mockAddSubCommand).toBeCalledTimes(2);
    expect(mockAddSubCommand).toHaveBeenNthCalledWith(1, mockChangeCmd);
    expect(mockAddSubCommand).toHaveBeenNthCalledWith(2, mockChangeCmd);
    expect(mockAddCommandToHistory).toBeCalledTimes(1);
    expect(mockAddCommandToHistory).toHaveBeenLastCalledWith(mockBatchCommand);
    expect(mockReRenderImageSymbol).toBeCalledTimes(1);
    expect(mockReRenderImageSymbol).toHaveBeenLastCalledWith(expect.anything(), { force: true });
    expect(getByText('initColor: #AAAAFF')).toMatchSnapshot();
  });
});
