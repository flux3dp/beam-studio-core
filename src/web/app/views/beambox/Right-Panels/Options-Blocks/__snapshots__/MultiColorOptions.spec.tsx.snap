// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MultiColorOptions editing color in use element should work 1`] = `
<p>
  initColor: 
  #AAAAFF
</p>
`;

exports[`test MultiColorOptions should render correctly 1`] = `
<div>
  <div
    class="block"
  >
    <div
      class="label"
    >
      Colors
    </div>
    <div
      class="controls"
    >
      <div>
        Mock ColorPicker
        <p>
          allowClear: 
          f
        </p>
        <p>
          initColor: 
          #AAFFFF
        </p>
        <p>
          triggerType: 
        </p>
        <button
          type="button"
        >
          onChange
        </button>
      </div>
      <div>
        Mock ColorPicker
        <p>
          allowClear: 
          f
        </p>
        <p>
          initColor: 
          #FFAAFF
        </p>
        <p>
          triggerType: 
        </p>
        <button
          type="button"
        >
          onChange
        </button>
      </div>
    </div>
  </div>
</div>
`;
